src/logging-and-metrics.interceptor.ts

```diff
@@ -0,0 +1,203 @@
+// Copyright © 2022 Kaleido, Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import {
+  CallHandler,
+  ExecutionContext,
+  HttpException,
+  Injectable,
+  Logger,
+  Module,
+  NestInterceptor,
+} from '@nestjs/common';
+import { Request, Response } from 'express';
+import { Observable, throwError } from 'rxjs';
+import { catchError, tap } from 'rxjs/operators';
+import { FFRequestIDHeader } from './request-context/constants';
+import { Counter, Histogram, Gauge } from 'prom-client';
+import {
+  InjectMetric,
+  makeCounterProvider,
+  makeGaugeProvider,
+  makeHistogramProvider,
+} from '@willsoto/nestjs-prometheus';
+
+const METRICS_ROUTE = '/metrics';
+
+export const MetricProviders = [
+  makeCounterProvider({
+    name: 'ff_apiserver_rest_requests_total',
+    help: 'Total REST API requests handled by the service',
+    labelNames: ['method', 'route', 'code'],
+  }),
+  makeHistogramProvider({
+    name: 'ff_apiserver_rest_request_size_bytes',
+    help: 'Size of REST API requests in bytes',
+    labelNames: ['method', 'route', 'code'],
+  }),
+  makeHistogramProvider({
+    name: 'ff_apiserver_rest_request_duration_seconds',
+    help: 'Duration of REST API requests',
+    labelNames: ['method', 'route', 'code'],
+  }),
+  makeGaugeProvider({
+    name: 'ff_event_batch_size',
+    help: 'Size of event batches delivered from blockchain connector',
+  }),
+  makeHistogramProvider({
+    name: 'ff_event_batch_interval_ms',
+    help: 'Interval between the arrival of batches of events, in milliseconds',
+    buckets: [250, 500, 750, 1000, 1250, 1500, 1750, 2000, 2500],
+  }),
+  makeHistogramProvider({
+    name: 'ff_event_batch_ack_interval_ms',
+    help: 'Interval between the arrival of batch acknowledgements, in milliseconds',
+    buckets: [250, 500, 750, 1000, 1250, 1500, 1750, 2000, 2250, 2500, 2750, 3000],
+  }),
+  makeCounterProvider({
+    name: 'ff_blockchain_calls_total',
+    help: 'Total calls to the blockchain connector',
+    labelNames: ['type'],
+  }),
+];
+
+@Module({})
+@Injectable()
+export class LoggingAndMetricsInterceptor implements NestInterceptor {
```

Comment:
It feels like this could still be a separate interceptor from `RequestLoggingInterceptor`. That one is shared across most of our Nest projects, whereas this one feels quite specific to FireFly token connectors.

If there's a benefit to combining these two interceptors, I'm not fundamentally opposed; just seemed like two different tasks.

Reply:
Personally it felt to me like they were both useful/common things to want to do as part of request interception, so bundled them together for that reason. I'm happy to split them apart if you think that would be best for the codebase, happy for you to take a call on it.